### PR TITLE
add new fields for borrowing in flowcontrol api

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Adds the new fields for borrowing support in flowcontrol API:
- https://github.com/kubernetes/enhancements/pull/3391 
- https://github.com/kubernetes/enhancements/pull/3479

#### Which issue(s) this PR fixes:
Fixes #

#### Does this PR introduce a user-facing change?
```release-note
Priority and Fairness has introduced a new feature called Borrowing that allows a priority level to borrow
a number of seats from other priority level(s). The cluster operator can enable borrowing for a certain 
priority level configuration object via the two newly introduced fields 'lendablePercent', and
`borrowingLimitPercent` located under the 'spec.limited' field of the designated priority level.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go [hunk 1]
```diff
@@ -438,6 +438,30 @@ type LimitedPriorityLevelConfiguration struct {
 
 	// `limitResponse` indicates what to do with requests that can not be executed right now
 	LimitResponse LimitResponse `json:"limitResponse,omitempty" protobuf:"bytes,2,opt,name=limitResponse"`
+
```
Comment (tkashem) [inline staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go:441]: @MikeSpreitzer I copied it from the KEPs, can you please review these two fields? (I am working on fixing the other errors)

File: staging/src/k8s.io/api/flowcontrol/v1beta3/types.go [hunk 2]
```diff
@@ -472,6 +472,30 @@ type LimitedPriorityLevelConfiguration struct {
...
+	// LendableCL(i) = round( NominalCL(i) * lendablePercent(i)/100.0 )
+	//
+	// +optional
+	LendablePercent int32 `json:"lendablePercent" protobuf:"varint,3,opt,name=lendablePercent"`
+
+	// `borrowingLimitPercent`, if present, specifies a limit on how many seats
+	// this priority level can borrow from other priority levels.  The limit
+	// is known as this level's BorrowingConcurrencyLimit (BorrowingCL) and
+	// is a limit on the total number of seats that this level may borrow
+	// at any one time.  When this field is non-nil, it must hold a non-negative
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/api/flowcontrol/v1beta3/types.go:491]: This is probably a copy of words I wrote, but now that I read them here I see a problem.  The words here say this field holds a number of seats; the truth is that this field holds a ratio.  Perhaps words like the following would be better?

```
// `borrowingLimitPercent`, if present, configures a limit on how many seats
// this priority level can borrow from other priority levels.  The limit
// is known as this level's BorrowingConcurrencyLimit (BorrowingCL) and
// is a limit on the total number of seats that this level may borrow
// at any one time.
// This field holds the ratio of that limit to the level's nominal concurrency limit.
// When this field is non-nil, it must hold a non-negative

Comment (MikeSpreitzer) [reply staging/src/k8s.io/api/flowcontrol/v1beta3/types.go:491]: The KEP change just merged.  Let's not worry about wording details there, just fix it here.

**Comment by tkashem** [comment]
> === FAIL: vendor/k8s.io/apimachinery/pkg/util/httpstream/spdy TestConnectionPings (0.09s)
>    connection_test.go:240: server: failed to send any pings (check logs)

/test pull-kubernetes-unit

**Review by MikeSpreitzer — APPROVED** [review]
/LGTM
Thanks!

**Review by BenTheElder — COMMENTED** [review]
Is the plan to merge this with the implementation?

This probably needs a release-note?

File: staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go [hunk 3]
```diff
@@ -438,6 +438,32 @@ type LimitedPriorityLevelConfiguration struct {
...
+	// `lendablePercent` prescribes the fraction of the level's NominalCL that
+	// can be borrowed by other priority levels.  This value of this
+	// field must be between 0 and 100, inclusive, and it defaults to 0.
+	// The number of seats that other levels can borrow from this level, known
+	// as this level's LendableConcurrencyLimit (LendableCL), is defined as follows.
+	//
+	// LendableCL(i) = round( NominalCL(i) * lendablePercent(i)/100.0 )
+	//
+	// +optional
+	LendablePercent int32 `json:"lendablePercent,omitempty" protobuf:"varint,3,opt,name=lendablePercent"`
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go:451]: I have no objection to making this a pointer.  I was only thinking about continuity with the current behavior when I said the default should be zero.

**Comment by tkashem** [comment]
i will let you know when it's ready for another pass, working through some test failures

**Comment by liggitt** [comment]
> @BenTheElder , to answer your question at [#112830 (review)](https://github.com/kubernetes/kubernetes/pull/112830#pullrequestreview-1141169209) : we plan to merge the API update before merging the implementation.

we avoid doing that so we don't accidentally release an API with inert fields that later spring to life. Once this has lgtm/approve, stack the implementation PR on top of it, review the implementation, and we can traffic direct them in

File: pkg/apis/flowcontrol/types.go [hunk 4]
```diff
@@ -416,6 +416,32 @@ type LimitedPriorityLevelConfiguration struct {
...
+	LendablePercent int32
+
+	// `borrowingLimitPercent`, if present, configures a limit on how many
+	// seats this priority level can borrow from other priority levels.
+	// The limit is known as this level's BorrowingConcurrencyLimit
+	// (BorrowingCL) and is a limit on the total number of seats that this
+	// level may borrow at any one time.
+	// This field holds the ratio of that limit to the level's nominal
+	// concurrency limit. When this field is non-nil, it must hold a
+	// non-negative integer and the limit is calculated as follows.
```
[the remark this answers is not in the document]
Comment (tkashem) [reply pkg/apis/flowcontrol/types.go:438]: done

File: staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tkashem) [reply staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go]: made this a pointer, and defaulting to zero, please take a look.

**Comment by tkashem** [comment]
@BenTheElder @MikeSpreitzer this is ready for another pass, thanks!

**Review by BenTheElder — COMMENTED** [review]
LGTM

(EDIT: other than https://github.com/kubernetes/kubernetes/pull/112830#issuecomment-1285620021, also I'm still learning API Review :-))

**Comment by tkashem** [comment]
@BenTheElder thanks for your review! this is my first time as well with API changes, defaulting, conversion etc. I am learning a lot here.

File: staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go [hunk 5]
```diff
@@ -82,6 +82,12 @@ var (
 	}
 )
 
+func makeInt32Ptr(v int32) *int32 {
+	p := new(int32)
+	*p = v
+	return p
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go:88]:
```suggestion
	return &v
```

**Review by MikeSpreitzer — APPROVED** [review]
/lgtm

**Comment by MikeSpreitzer** [comment]
Here is the current plan/status for following steps to complete the implementation.

1. Update the config-producing controller to set the new fields.  This is already submitted, #113016 .
2. Define the constants for the symbolic constants that appear in the borrowing design.  Already submitted, #113189 .
3. Define/plumb the new metrics introduced for borrowing.
4. Implement the collection of the statistics that drive the borrowing.  Just submitted, #113222 .
5. Modify the APF config-consuming controller so that it has a set of variable concurrency limits, initialized from config but not periodically adjusted.
6. Add the periodic adjustments.

File: pkg/apis/flowcontrol/validation/validation.go [hunk 6]
```diff
@@ -388,6 +388,14 @@ func ValidateLimitedPriorityLevelConfiguration(lplc *flowcontrol.LimitedPriority
 		allErrs = append(allErrs, field.Invalid(fldPath.Child(getVersionedFieldNameForConcurrencyShares(requestGV)), lplc.NominalConcurrencyShares, "must be positive"))
 	}
 	allErrs = append(allErrs, ValidateLimitResponse(lplc.LimitResponse, fldPath.Child("limitResponse"))...)
+
+	if lplc.LendablePercent != nil && !(*lplc.LendablePercent >= 0 && *lplc.LendablePercent <= 100) {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("lendablePercent"), *lplc.LendablePercent, "must be between 0 and 100, inclusive"))
+	}
+	if lplc.BorrowingLimitPercent != nil && *lplc.BorrowingLimitPercent <= 0 {
```
Comment (liggitt) [inline pkg/apis/flowcontrol/validation/validation.go:395]: shouldn't `BorrowingLimitPercent == 0` be valid, to allow disabling borrowing? this treats it as invalid

File: pkg/apis/flowcontrol/validation/validation_test.go [hunk 7]
```diff
@@ -1434,3 +1442,97 @@ func TestValidateLimitedPriorityLevelConfiguration(t *testing.T) {
...
+			borrowingLimitPercent: makeInt32Ptr(1),
+			errExpected:           nil,
+		},
+		{
+			borrowingLimitPercent: makeInt32Ptr(100),
+			errExpected:           nil,
+		},
+		{
+			borrowingLimitPercent: makeInt32Ptr(0),
+			errExpected:           errBorrowingLimitPercentFn(0),
```
Comment (liggitt) [inline pkg/apis/flowcontrol/validation/validation_test.go:1512]: I expected this to be allowed, not rejected

File: staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go [hunk 8]
```diff
@@ -82,6 +82,12 @@ var (
 	}
 )
 
+func makeInt32Ptr(v int32) *int32 {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go:85]: nit: use `k8s.io/utils/pointer#Int32()` instead?

File: staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go [hunk 9]
```diff
@@ -82,6 +82,12 @@ var (
...
+	p := new(int32)
+	*p = v
+	return p
+}
+
 // Mandatory PriorityLevelConfiguration objects
 var (
 	MandatoryPriorityLevelConfigurationExempt = newPriorityLevelConfiguration(
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go:93]: I'm curious what prompted updating these, if there was a test that noticed/failed when these didn't match their round-tripped/defaulted versions?

**Comment by liggitt** [comment]
one comment on the borrowing validation, a couple nits/questions, lgtm otherwise

File: pkg/apis/flowcontrol/validation/validation.go (hunk 6 above)
Comment (tkashem) [reply pkg/apis/flowcontrol/validation/validation.go:395]: yes, i overlooked it, `0` should be allowed. I already have a test, just need to tweak the test to allow it.

File: staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go (hunk 9 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go:93]: yes, there is a case here:
https://github.com/kubernetes/kubernetes/blob/4d50d28777795b01195368b27964458d798248dd/pkg/apis/flowcontrol/validation/validation.go#L348-L356

when a new mandatory `PriorityLevelConfiguration` object is created, the validation logic ensures that the spec defined in the source code (not defaulted before this PR) matches the object being created.

The integration job was failing due to this issue. Do you think deep copying and defaulting the spec (in the source file) before the comparison is a more preferable approach here?

currently, we have a test that verifies that defaulting any bootstrap object in the source code does not change the spec:
https://github.com/kubernetes/kubernetes/pull/112830/commits/69f3626d74470c9bc80895af8f895046ed866960#diff-91b89583335fd122fb47dc639206dcace01bf7812d7db0ea7eca697b17b29e50R31-R59

File: staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go (hunk 8 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go]: i am using the util function now

File: staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go (hunk 9 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go]: > currently, we also have a test that verifies that defaulting bootstrap object(s) in the source code does not change the spec

That was what I was expecting. The check in API validation was surprising to me. I need to think about whether making API validation compare the object in the request to the compiled-in object introduces any issues with skewed servers. That doesn't block this PR though.

**Comment by liggitt** [comment]
API lgtm, will tag once implementation PRs are ready to merge as well

**Comment by MikeSpreitzer** [comment]
@liggitt and others: If we have to hold the API change until the implementation is complete, that puts a different stress on ordering than I expected.  To be clear, some of the implementation items can be merged before the API change.  I am thinking of the following.

- #113189 just adds non-public constants to an implementation package.
- #113222 adds some instrumentation.
- we do not yet have a PR for it, but the new metrics could be defined --- and the relevant subset could be connected to the instrumentation added in #113222 .

**Comment by MikeSpreitzer** [comment]
@liggit: I am going to create the remainder of the implementation today.  Some of it will have no dependency on the new API, and some _will_ depend on the new API.  Looking for your advice on how to wrangle this work into PRs.  I am assuming the stuff that has no dependency on the API goes in its own PR(s).  For the stuff that _does_ depend on the API, as mentioned we have factored it into a few pieces.  If I understand correctly, you want those pieces added into this PR.  I assume as one or more commits.  Would that be one commit with all the pieces, or one commit for each of the pieces?  If the latter: do we add each commit to the PR as that commit becomes ready, or collect them elsewhere and add them all in one push once they are all ready?

**Comment by liggitt** [comment]
normally one of the following two patterns is followed:
1. implementation commits get added to the API PR as subsequent commits, reviewed by implementation reviewers, then the whole things gets merged
2. implementation commits get stacked on top of the API PR commits in a new PR, reviewed by implementation reviewers, then the whole things gets merged, citing the already-done review of the API PR (if they are literally the same commit SHAs, github will even mark the API PR as merged when the implementation PR merges)

either way works, which path you take mostly depends on who will be doing the work and how you want to manage any updates (from rebases, etc) required to the PR(s)

**Comment by MikeSpreitzer** [comment]
@liggitt : thanks.  I think I will pursue the second approach, since it more clearly does not disturb the API review that has been done.

**Comment by MikeSpreitzer** [comment]
@liggitt : #113485 is the full PR.  It starts with the commits from this PR, but I had to rebase because there were some merge conflicts (in the implementation details) with intervening progress on `master`.  Do I need to do anything formal regarding API approval for #113485 ?

**Comment by liggitt** [comment]
> @liggitt : #113485 is the full PR. It starts with the commits from this PR, but I had to rebase because there were some merge conflicts (in the implementation details) with intervening progress on `master`. Do I need to do anything formal regarding API approval for #113485 ?

Just let the API reviewer from this PR (me) know the impl PR is ready.

For anyone following along, since the sha's were different, I rebased both this PR and that PR on `master` locally and diffed the API commits. The only changes were to these files:

```
5       5       staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go
117     0       staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default_test.go
```

the non-test diff was:

```diff
diff --git a/staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go b/staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go
index cbfaca84e9c..3859a54d1fe 100644
--- a/staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go
+++ b/staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go
@@ -170,7 +170,7 @@ var (
                        Type: flowcontrol.PriorityLevelEnablementLimited,
                        Limited: &flowcontrol.LimitedPriorityLevelConfiguration{
                                NominalConcurrencyShares: 30,
-                               LendablePercent:          pointer.Int32(0),
+                               LendablePercent:          pointer.Int32(33),
                                LimitResponse: flowcontrol.LimitResponse{
                                        Type: flowcontrol.LimitResponseTypeQueue,
                                        Queuing: &flowcontrol.QueuingConfiguration{
@@ -187,7 +187,7 @@ var (
                        Type: flowcontrol.PriorityLevelEnablementLimited,
                        Limited: &flowcontrol.LimitedPriorityLevelConfiguration{
                                NominalConcurrencyShares: 40,
-                               LendablePercent:          pointer.Int32(0),
+                               LendablePercent:          pointer.Int32(25),
                                LimitResponse: flowcontrol.LimitResponse{
                                        Type: flowcontrol.LimitResponseTypeQueue,
                                        Queuing: &flowcontrol.QueuingConfiguration{
@@ -223,7 +223,7 @@ var (
                        Type: flowcontrol.PriorityLevelEnablementLimited,
                        Limited: &flowcontrol.LimitedPriorityLevelConfiguration{
                                NominalConcurrencyShares: 40,
-                               LendablePercent:          pointer.Int32(0),
+                               LendablePercent:          pointer.Int32(50),
                                LimitResponse: flowcontrol.LimitResponse{
                                        Type: flowcontrol.LimitResponseTypeQueue,
                                        Queuing: &flowcontrol.QueuingConfiguration{
@@ -241,7 +241,7 @@ var (
                        Type: flowcontrol.PriorityLevelEnablementLimited,
                        Limited: &flowcontrol.LimitedPriorityLevelConfiguration{
                                NominalConcurrencyShares: 100,
-                               LendablePercent:          pointer.Int32(0),
+                               LendablePercent:          pointer.Int32(90),
                                LimitResponse: flowcontrol.LimitResponse{
                                        Type: flowcontrol.LimitResponseTypeQueue,
                                        Queuing: &flowcontrol.QueuingConfiguration{
@@ -259,7 +259,7 @@ var (
                        Type: flowcontrol.PriorityLevelEnablementLimited,
                        Limited: &flowcontrol.LimitedPriorityLevelConfiguration{
                                NominalConcurrencyShares: 20,
-                               LendablePercent:          pointer.Int32(0),
+                               LendablePercent:          pointer.Int32(50),
                                LimitResponse: flowcontrol.LimitResponse{
                                        Type: flowcontrol.LimitResponseTypeQueue,
                                        Queuing: &flowcontrol.QueuingConfiguration{
```

I'll defer to the impl reviewer for the details of the default policies, and all the other API additions/changes are identical to what was reviewed here, so I'll transfer the "API approved" comment over to #113485

**Comment by MikeSpreitzer** [comment]
This was rebased and included in #113485 .  Should be closed now, but I do not see how to do that.

**Pull request closed without merging**